### PR TITLE
BigBench-Hard dataset added (IO and COT Methods). Readme updated. Utils file created, Results code refactored.

### PR DESCRIPTION
**Pull request by AjayDextrous**

### CODE REVIEW

File: examples/bigbench_hard/bigbench_hard.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        current_state = thought.state["current"]
+        scores[current_state] = scores.get(current_state, 0) + 1
+    logging.info("return_most_frequent_answer: scores: {}".format(scores))
+    frequent_answer = max(scores, key=scores.get)
+    for thought in thoughts:
+        thought.score = scores[thought.state["current"]]
+        if ignore_empty_answers:
+            if thought.state is None:
+                thought.score = 0
+    return thoughts
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py:35]: smart to score by frequency! I just did 1 if majority else 0. Will add this too (or we could make this scoring function part of the score operation in an extra branch.. I think we will need it more often..)

File: project_utils.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import json
```
Comment (fjfricke) [inline project_utils.py:1]: nice helpful file!

File: examples/bigbench_hard/bigbench_hard.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        full_prompt: str = ""
+        with open(prompt_path, "r") as f:
+            for _ in range(2):  # to skip canary warning
+                f.readline()
+            prompt: str = f.readline().rstrip("\n")  # prompt is always on the 3rd line for each task
+            examples: [str] = f.read().lstrip("\n").split("\n\n")  # the rest of the file constitutes the example
+            answers: [str] = []
+            for example in examples:
+                answers.append(example.split("So the answer is ")[-1].removesuffix("."))  # extract answer from example
+            full_examples: [str] = []
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py]: please change them from str to List[str].
```suggestion
            examples: List[str] = f.read().lstrip("\n").split("\n\n")  # the rest of the file constitutes the example
            answers: List[str] = []
            for example in examples:
                answers.append(example.split("So the answer is ")[-1].removesuffix("."))  # extract answer from example
            full_examples: List[str] = []
```

File: examples/bigbench_hard/bigbench_hard.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    """
+    Generates the Graph of Operations for the IO method.
+
+    :return: Graph of Operations
+    :rtype: GraphOfOperations
+    """
+    operations_graph = operations.GraphOfOperations()
+
+    operations_graph.append_operation(operations.Generate(1, 1))
+    operations_graph.append_operation(operations.GroundTruth(test_answer))
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py:203]: I think for plotting, you have to add a final scoring method.. we should discuss this

File: examples/bigbench_hard/bigbench_hard.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    """
+    Generates the Graph of Operations for the COT method.
+
+    :return: Graph of Operations
+    :rtype: GraphOfOperations
+    """
+    operations_graph = operations.GraphOfOperations()
+
+    operations_graph.append_operation(operations.Generate(1, 1))
+    operations_graph.append_operation(operations.GroundTruth(test_answer))
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py:218]: same

File: examples/bigbench_hard/bigbench_hard.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py:233]: same

File: examples/bigbench_hard/bigbench_hard.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    :rtype: GraphOfOperations
+    """
+    num_branches = 5
+    operations_graph = operations.GraphOfOperations()
+
+    generate_operation = operations.Generate(1, num_branches)
+    operations_graph.append_operation(generate_operation)
+    operations_graph.append_operation(
+        operations.Selector(selector=score_answers_by_frequency))  # have to do this less than ideal implementation
+    # due to issues with existing scoring function, might be better to refactor
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py:252]: I did it the same way. I think we can add a new branch that adds frequency scoring directly to the operation

File: examples/bigbench_hard/bigbench_hard.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+     """
+    raise NotImplementedError("Not Implemented")
+
+
+def run(
+        data_ids: List[int],
+        methods: List[Callable[[], operations.GraphOfOperations]],
+        budget: float,
+        lm_name: str,
+        tasks: [str] = [],
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py]:
```suggestion
        tasks: List[str] = [],
```

File: examples/bigbench_hard/bigbench_hard.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+     Generates the Graph of Operations for the TOT method.
+
+     :return: Graph of Operations
+     :rtype: GraphOfOperations
+     """
+    raise NotImplementedError("Not Implemented")
+
+
+def run(
+        data_ids: List[int],
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py]: you should implement using the data_ids somehow or in some other way make sure that you can select the number of samples

Comment (fjfricke) [reply examples/bigbench_hard/bigbench_hard.py]: at the moment your code runs on all examples. This is very costly for testing your code

File: examples/bigbench_hard/bigbench_hard.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+                    output_path: Path = method_results_dir / f"{id}.json"
+                    executor.output_graph(str(output_path))
+                    budget -= lm.cost
+
+    return orig_budget - budget
+
+
+if __name__ == "__main__":
+    budget = 30
+    samples = [item for item in range(5)]
```
Comment (fjfricke) [inline examples/bigbench_hard/bigbench_hard.py]: meaning `samples` is not used

**Comment by fjfricke** [comment]
In general it runs and works. Plotting could also be easily implemented on a task basis

File: examples/bigbench_hard/bigbench_hard.py (hunk 1 above)
Comment (AjayDextrous) [reply examples/bigbench_hard/bigbench_hard.py:35]: Yes, Let's implement a `RelativeScorer` in our next iteration

**Review by Jiaqi7han — APPROVED** [review]

**Pull request merged**